zebra-state/src/service/read/difficulty.rs

```diff
@@ -87,42 +88,55 @@ pub fn solution_rate(
     start_hash: Hash,
 ) -> Option<u128> {
-    // Take 1 extra block for calculating the number of seconds between when mining on the first block likely started.
-    // The work for the last block in this iterator is not added to `total_work`.
-    let mut block_iter = any_ancestor_blocks(non_finalized_state, db, start_hash)
-        .take(num_blocks.checked_add(1).unwrap_or(num_blocks))
-        .peekable();
-
-    let get_work = |block: Arc<Block>| {
-        block
-            .header
+    // Take 1 extra header for calculating the number of seconds between when mining on the first
+    // block likely started. The work for the extra header is not added to `total_work`.
+    //
+    // Since we can't take more headers than are actually in the chain, this automatically limits
+    // `num_blocks` to the chain length, like `zcashd` does.
+    let mut header_iter =
+        any_chain_ancestor_iter::<block::Header>(non_finalized_state, db, start_hash)
+            .take(num_blocks.checked_add(1).unwrap_or(num_blocks))
+            .peekable();
+
+    let get_work = |header: &block::Header| {
+        header
             .difficulty_threshold
             .to_work()
             .expect("work has already been validated")
     };
 
-    let block = block_iter.next()?;
-    let last_block_time = block.header.time;
-
-    let mut total_work: PartialCumulativeWork = get_work(block).into();
-
-    loop {
-        // Return `None` if the iterator doesn't yield a second item.
-        let block = block_iter.next()?;
-
-        if block_iter.peek().is_some() {
-            // Add the block's work to `total_work` if it's not the last item in the iterator.
-            // The last item in the iterator is only used to estimate when mining on the first block
-            // in the window of `num_blocks` likely started.
-            total_work += get_work(block);
-        } else {
-            let first_block_time = block.header.time;
-            let duration_between_first_and_last_block = last_block_time - first_block_time;
-            return Some(
-                total_work.as_u128() / duration_between_first_and_last_block.num_seconds() as u128,
-            );
-        }
+    // If there are no blocks in the range, we can't return a useful result.
+    let last_header = header_iter.peek()?;
+
+    // Initialize the cumulative variables.
+    let mut min_time = last_header.time;
+    let mut max_time = last_header.time;
+
+    let mut last_work = Work::zero();
+    let mut total_work = PartialCumulativeWork::zero();
+
+    for header in header_iter {
+        min_time = min_time.min(header.time);
+        max_time = max_time.max(header.time);
+
+        last_work = get_work(&header);
+        total_work += last_work;
     }
+
+    // We added an extra header so we could estimate when mining on the first block
+    // in the window of `num_blocks` likely started. But we don't want to add the work
+    // for that header.
+    total_work -= last_work;
```

Comment:
Nitpick/Optional:

(also remove `Work` import if applying this)

```suggestion
    let mut total_work = PartialCumulativeWork::zero();
    let mut next_total_work = PartialCumulativeWork::zero();

    for header in header_iter {
        min_time = min_time.min(header.time);
        max_time = max_time.max(header.time);

        total_work = next_total_work;

        // We added an extra header so we could estimate when mining on the first block
        // in the window of `num_blocks` likely started. But we don't want to add the work
        // for that header.
        next_total_work += get_work(&header);
    }
```

Reply:
Thanks but if this is needed please make another PR, I'd like to get this PR merged to avoid merge conflicts.